### DISCUSSION TITLE
"Pixel editor" not following exact number of pixels

### DISCUSSION THREAD

**Discussion by bryanseger** [sn4k3/UVtools#834] [General]
When using the pixel editor, for a square brush shape for example, if I type in "5" pixels for the brush diameter, the output isn't always a square that is exactly 5x5 pixels. 

If I do a 1 pixel, I get exactly that, but from 2-5 pixels, I get a square that is 2x2, if I do 6-7 pixels, I get a square that is 4x4. 8-9 pixels gives me a square of 6x6, etc.

If this could be fixed in the next update, that would be greatly appreciated! The attached pic is just an example of what I was testing just now.

![image](https://github.com/sn4k3/UVtools/assets/4185392/a265a738-6938-491c-af54-53ba65a21285)

**Comment by sn4k3** [comment]
The render preview is different from applied pixels. After apply you should get the number more correctly, depending on shape it can have an error. Also when applied with AntiAlias the grey pixels should be counted.
But I will look at code to see if I can improve this situation.

**Comment by bryanseger** [comment]
I saw that the last update mentioned a few things about the Pixel Editor, so I gave it another try. 

First, I do see what you mean that the preview is different than the actually applied pixels. This can make it cumbersome when we're trying to dial in specific pixels to be on or off.

I'm on v4.2.1 right now and just made the below. I tested all 3 "line type" options for the 'square' brush shape and pixel counts 1 through 10. We generally don't use anti-aliasing, but I included it here just to represent all options in one image. Only brush diameters of 1 and 3 pixels are correct. 

I'm wondering if this can be further improved on. We've been "making it work" and getting by, but if this could be tightened up, that would increase our productivity big time! Thanks Tiago for looking into this.

![image](https://github.com/sn4k3/UVtools/assets/4185392/5481a826-002b-49f5-86fe-c0143bd1248b)

**Reply by sn4k3** [reply]
The updates regarding pixel editor was about profiles and UI improvements. The code of shapes still unchanged.

The "problem" is the code of the shapes are in fact math code to produce regular polygons using Sin and Cos from a center point and diameter with a possibility of a rotation angle. Due the nature of the code it will never be accurate, as brush diameter is in fact the diameter of the shape from it center and not side to side pixels, and rounding problems of the pixels made the shapes to have "unexpected" sizes so this affect all shapes, this is why you see in most cases is missing a row and cell. I have tried to use the openCV RotatedRect but same "problem" happens as it also use math to define the vertex. So to have a perfect square I would have to use raw Rectangle and ignore the angle feature. However I will try to make an exception for square and try to compensate the missing pixels somehow.

Here you can see what's going on:

https://github.com/sn4k3/UVtools/blob/115ee060889f0425435cbedda334609c0783747d/UVtools.Core/Extensions/DrawingExtensions.cs#L41-L112

**Reply by bryanseger** [reply]
I understand now why that's happening, and I appreciate the explanation. 

Yes, if you're able to make an exception for the square brush, that would be an amazing time saver for our team. Thanks again!

**Reply by sn4k3** [reply]
If you want to use 100% accurate shape, use the circle brush, however it not optimal for small sizes.

**Reply by sn4k3** [reply]
Check latest release. Should be better now.

**Reply by bryanseger** [reply]
Yes, much better! Thanks Tiago!
